sqlcapture/capture.go

```diff
@@ -832,5 +835,5 @@ type StreamID = string
 
 // JoinStreamID combines a namespace and a stream name into a dotted name like "public.foo_table".
 func JoinStreamID(namespace, stream string) StreamID {
-	return strings.ToLower(namespace + "." + stream)
+	return namespace + "." + stream
```

Comment:
To ask the obvious question: Are we certain changing this doesn't break some assumption elsewhere in a SQL capture about the `StreamID` being lower-case? Spot checking I can't find anything obvious, but I didn't do a deep dive.  I'm assuming the reason for making this change is that Oracle names default to uppercase, and I see above that we have to do an extra check for the capitalized `"WATERMARK"`.

I'm wondering if the Oracle capture could set its `event.After` to have `"watermark"` as a lower case, and then we could keep the Stream ID creation unchanged?

Reply:
Yes this is to accommodate Oracle names being uppercase by default... I'm also not completely easy with this change.

Is there a reason for us to lowercase all streamIDs though, other than to avoid duplicate streamIDs (which I think should be forbidden before sqlcapture gets involved anyway)? @willdonnelly perhaps you may know something about this?

Reply:
There used to be good reason, but we've been slowly working towards fixing that, so we're almost at the point where they can be made case-sensitive now.

The last major roadblock, IIRC, was resolved in the state key migration. It'd be a good idea to double check, but I think that with that changed we're no longer persisting StreamID values across connector runs, so it might be as simply as just removing it. We'd want to double-check pretty thoroughly that it doesn't break anything in production before rolling that out, but it might be time to bite that bullet.

Also if/when we make it case sensitive I'd really like to try redefining the StreamID type as something like `struct { Schema, Table string }` so it actually has proper type safety and isn't just an alias for a string.

Reply:
Okay, I've finally had some time to read through the latest code and now I'm even more confused why exactly the stream IDs need to be case sensitive for Oracle specifically.

As far as I can tell no stream ID is ever being split back into schema or table name components, so additional normalization should be fine unless we're specifically worried about capturing two tables with identical-except-capitalization names. Is that the only issue we're concerned with here, and if so is there some reason that's a bigger issue with Oracle than other databases?

Reply:
@willdonnelly The main issue arises here:
https://github.com/estuary/connectors/blob/main/sqlcapture/main.go#L167-L170

The bindings are mapped keyed on the StreamID, but during discovery I want the table names to retain their original casing (uppercase by default in case of Oracle). If I do discovery with uppercase names, then I end up with this error due to the discrepancy:

```
error updating capture state: table "flow_test_logminer.t95911555" is a configured binding of this capture, but doesn't exist or isn't visible with current permissions
```

Reply:
I'm afraid I'm still confused, what exactly do you mean by "retain their original casing"?

The original schema and table names are present in the `DiscoveryInfo.Schema` and `DiscoveryInfo.Name` properties, which then become the resource spec `Namespace` and `Stream` properties. All of the generic discovery logic should be based on that and not care at all about the stream ID used for a table. I've just double-checked the relevant discovery code to be sure and I still see nowhere that the stream IDs should be making it into the actual discovered resources or schemas, but if that is happening it's a bug.

The bit of logic you linked should also work correctly if stream IDs are normalized to lowercase. The appropriately-capitalized schema and table name come from the resource spec, and a lowercase stream ID is generated from them and used to look up a value from the discovery output. The key of the discovery map is also a lowercased stream ID generated from the appropriately-capitalized schema/table names that discovery just found. Where does the problem come from?

Like, in principle we ought to be able to just come along one day and swap out the implementation of `JoinStreamID` for something that hashes the names and everything else would keep going without caring. The actual representation of these IDs isn't supposed to matter and if it does that implies a bug somewhere else.

So the only thing that should be broken by normalizing stream IDs to lowercase is that we can't support tables with the exact same name but differently capitalized. I'd like to make them case sensitive at some point so that stops being an issue for all SQL CDC connectors, but it doesn't sound like that's the problem you're trying to avoid here? The fact remains that **if anything else is broken by this, the IDs are being used incorrectly somewhere**.

Reply:
@willdonnelly okay that's a good perspective on how streamID is supposed to be handled. I dug deeper and I found what was causing this issue:

1. The `DiscoverTables` method that is to be implemented by each connector is expected to have `StreamID` keys, and in many places in `sqlcapture` we use an `StreamID` to check for existence of a table
3. In case of Oracle, the `DiscoverTables` is using `sqlcapture.JoinStreamID`, so the discover tables is keyed on `StreamID`s as expected
4. However, in the case of the watermarks table, the check is done like so: https://github.com/estuary/connectors/blob/main/sqlcapture/capture.go#L174-L177 which expects the value returned by `WatermarksTable()` method to match one of the keys in the discovery output. In this case, the Oracle connector was outputting an all-uppercase `<USER>.FLOW_WATERMARKS` table name, which would not match the lowercased keys of discovery.
5. This was causing the following error:
```
error activating replication for watermarks table %q: table was not observed by autodiscovery
```
6. The issue this arises is that if the user enters a watermarks table that is case-sensitive, we will fail with the same error, since the case-sensitive table name will not match its lowercased counterpart. This seems to affect all connectors, but this is probably more common among Oracle users to want to enter their watermarks table in uppercase. So to _workaround_ this limitation we will need to also lowercase the `WatermarksTable()` output of Oracle

So I think the issue here is that we are using `JoinStreamID` for keys of the discovery, but not for checking the existence of `WatermarksTable()`, so there is a possibility for inconsistency here.

Reply:
Oh, interesting. Yeah, it makes sense it'd be watermarks table handling that causes the problem there, since that's the one spot we're exposing something like a stream ID directly to the user.

I have a vague notion of how to fix this (basically recognizing that the watermarks table name is its own thing which is not necessarily a stream ID, and modifying the appropriate interface bits to split `cfg.Advanced.WatermarksTable` explicitly and then feed the schema/table name components into JoinStreamID) but I'll have to sit down and try it to be sure if that would work.

Reply:
I think for now I'll take this limitation of case-sensitive watermarks tables, put a `strings.ToLower` around the watermarks table configuration of Oracle. We should definitely revisit this though, I think your idea sounds good both for addressing the watermarks table, and perhaps to generally move from the current `StreamID` into an actual struct

Reply:
Yeah, that's basically the end goal of making sure that nothing else cares about the internal representation of a stream ID -- if it doesn't matter then we can just swap it over to being a struct and nothing will care, at which point we can be less vigilant against guarding against any sort of "misuse" of the values.

But none of the other connectors handle watermark table names correctly at this spot either, so just lowercasing seems fine for now, and I'll put the same "split using a custom function and then recombine with JoinStreamID" behavior in the Oracle connector when I do the others.